File: JavaScript/008_bank.js
const prompt = require("prompt-sync")({sigint:true});

class BankAccount {
    constructor(balance) {
        this._balance = balance;
    }

    balance() {
        let formattedBalance = this._balance.toFixed(2);
        console.log(`Your current balance is R${formattedBalance}`);
    }

    isDigit(amount) {
        return /^\d+$/.test(amount);
    }

    deposit() {
        let deposit = Number(prompt("How much do you want to deposit: "));
        if (this.isDigit(deposit)) {
            this._balance = this._balance + deposit;
        } else {
            console.log("An exception occurred. Please try again.");
        }
    }

    withdraw() {
        let withdrawal = Number(prompt("How much do you want to deposit: "));
        if (this.isDigit(withdrawal)) {
            if (withdrawal < this._balance) {
            this._balance = this._balance - withdrawal;
            } else {console.log("You have insufficient balance.")}
        } else {
            console.log("An exception occurred. Please try again.");
        }
    }
}

function main() {
    let balanceAmount = 0.00;
    let bank = new BankAccount(balanceAmount);
    let bankOn = true;

    while (bankOn) {
        bank.balance();
        console.log("1. Deposit"+"\n"+
                    "2. Withdraw"+"\n"+
                    "3. Quit");
        let status = Number(prompt("what do you want to do.(Select a number): "));
        if (bank.isDigit(status)) {
            if (status == 1) {
                bank.deposit();
            }
            if (status == 2) {
                bank.withdraw();
            }
            if (status == 3) {
                bank.balance();
                console.log("GoodBye. See you soon.");
                bankOn = false;
            }
        } else {
            console.log("Select a valid choice.");
        }
    }
}

main()

File: Python/008_bank.py
# Task 9: Implement a simple banking system. Create a BankAccount class that
# allows for deposits and withdrawals, and keeps track of the current balance. 
# Ensure that the account cannot go below $0.

class BankAccount:

    # initialized balance
    def __init__(self, balance):
        self.balance = balance

    def balance_def(self):
        number = format(self.balance, ".2f")
        print(f"Your current balance is R{number}")

    # deposit adds amount to balance
    def deposit(self):
        try:
          deposit = float(input("How much do you want to deposit: R"))
          self.balance = self.balance + deposit    
        except:
          print('An exception occurred. Please try again.')
    
    # withdraw subtracts amount from balance
    # can go beyond 0
    def withdraw(self):
        try:
            amountWithdraw = float(input("How much do you want to withdraw: R"))
            if (self.balance - amountWithdraw) > 0:
                self.balance = self.balance - amountWithdraw
            else:
                print("You have insufficient balance.")
        except:
          print('An exception occurred. Please try again.')


if __name__ == "__main__":

    balance = 0.00
    bank = BankAccount(balance)
    bankOn = True

    # Keeps the banks running until otherwise
    # used try to catch all errors
    
    while bankOn:
        bank.balance_def()
        print("1. Deposit"+"\n"
              "2. Withdraw"+"\n"
              "3. Quit")
        try:
            status = int(input("what do you want to do.(Select a number): "))
            if status == 1:
               bank.deposit()
            elif status == 2:
               bank.withdraw()
            elif status == 3:
               bank.balance_def()
               print("Goodbye")
               bankOn = False
            else:
               print("Select a valid choice (Select a number)")
               continue
        except:
          print('An exception occurred. Please try again.')
        
        

